### PR TITLE
Redis-ha fixes

### PR DESCRIPTION
**Pull request by ianmaddox**
Streamlining init process to only require one kind of Redis server in addition to the Sentinels. The Redis instances deterministically pick their own master now.
Depends on quay.io/ianmaddox/redis:4.0.6r0
Tightening outage detection
Adding labels for diagnostics
Using Redis 4.0.6
Making Sentinels responsible for moving Redis boxes in/out of master LB on promotion/demotion using labels
~~Chart version now 1.1.0~~
Chart version now 2.0.0

*****************************************************************************************

Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the 
history. This will make it easier to identify new changes. The PR will be squashed 
anyways when it is merged. Thanks.

*****************************************************************************************

### CODE REVIEW

File: stable/redis-ha/README.md [hunk 1]
```diff
@@ -47,7 +47,7 @@ The following tables lists the configurable parameters of the Redis chart and th
 
 | Parameter                        | Description                                           | Default                                                   |
 | -------------------------------- | ----------------------------------------------------- | --------------------------------------------------------- |
-| `redis_image`                    | Redis image                                           | `quay.io/smile/redis:4.0.2`                               |
+| `redis_image`                    | Redis image                                           | `quay.io/ianmaddox/redis:4.0.6`                           |
```
Comment (timstoop) [inline stable/redis-ha/README.md]: Why use a custom image when there's an official upstream image? Do you add anything to that image that's not present in the custom one? Could you document this as well?

File: stable/redis-ha/templates/redis-master-appliance-vpc.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: redis-master-appliance-vpc
```
Comment (timstoop) [inline stable/redis-ha/templates/redis-master-appliance-vpc.yaml]: It's customary in the official charts to create all resources with the fullname variable in front of it.

File: stable/redis-ha/templates/redis-sentinel-deployment.yaml [hunk 3]
```diff
@@ -3,23 +3,14 @@ kind: Deployment
...
-  replicas: {{ .Values.replicas.sentinel }}
+  replicas: {{ .Values.replicas.sentinels }}
   template:
     metadata:
       labels:
-        name: redis-sentinel
-        redis-sentinel: "true"
-        role: sentinel
+        app: "redis-ha"
+        name: "redis-sentinel"
```
Comment (timstoop) [inline stable/redis-ha/templates/redis-sentinel-deployment.yaml]: Same as previous comment, it's customary to have the fullname prefixing names of resources, as that allows one to have multiple setups.

File: stable/redis-ha/templates/redis-sentinel-deployment.yaml [hunk 4]
```diff
@@ -3,23 +3,14 @@ kind: Deployment
...
 {{ include "labels.standard" . | indent 8 }}
     spec:
-      {{- if .Values.nodeSelector }}
```
Comment (timstoop) [inline stable/redis-ha/templates/redis-sentinel-deployment.yaml]: You're removing an entire feature of the chart here by removing the nodeSelector parts. Why do that? Some people will like the flexibility of being able to select the nodes on which the Pods should run.

**Comment by viglesiasce** [comment]
cc @smileisak for review.

Looks like this makes backwards incompatible changes so would need to be a major version bump.

File: stable/redis-ha/README.md (hunk 1 above)
Comment (ianmaddox) [reply stable/redis-ha/README.md]: The custom image is part of core functionality for this chart. I updated the readme to cover the requirements.

File: stable/redis-ha/templates/redis-slave-appliance-vpc.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+spec:
+  ports:
+  - port: 6379
+    protocol: TCP
+    targetPort: 6379
+  selector:
+    app: "redis-ha"
+    redis-node: "true"
+    redis-role: "slave"
+  type: LoadBalancer
```
Comment (smileisak) [inline stable/redis-ha/templates/redis-slave-appliance-vpc.yaml]: There is a reason to use a `LoadBalancer` service type ?

File: stable/redis-ha/values.yaml [hunk 6]
```diff
@@ -66,9 +53,8 @@ persistentVolume:
   ##
   subPath: ""
 ## Redis image version
-redis_image: quay.io/smile/redis:4.0.2
+redis_image: quay.io/ianmaddox/redis-sentinel:4.0.6
```
Comment (smileisak) [inline stable/redis-ha/values.yaml]: Can you add image source files to understand what you did ?

File: stable/redis-ha/templates/redis-slave-statefulset.yaml [hunk 7]
```diff
@@ -1,44 +0,0 @@
-apiVersion: apps/v1beta1
```
Comment (smileisak) [inline stable/redis-ha/templates/redis-slave-statefulset.yaml:1]: Why did you delete this statefulset ?

**Review by smileisak — CHANGES_REQUESTED** [review]
@ianmaddox can you explain what you did to the docker image to have this radical change ?

File: stable/redis-ha/templates/redis-master-appliance-vpc.yaml (hunk 2 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-master-appliance-vpc.yaml]: Switched to dynamic names.

File: stable/redis-ha/templates/redis-sentinel-deployment.yaml (hunk 3 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-sentinel-deployment.yaml]: Same as above

File: stable/redis-ha/templates/redis-sentinel-deployment.yaml (hunk 4 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-sentinel-deployment.yaml]: Looks like this was added after the fork. I reintroduced the changes.

File: stable/redis-ha/templates/redis-slave-appliance-vpc.yaml (hunk 5 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-slave-appliance-vpc.yaml]: Hey Ismail. Hey From a high level, there are only two types of pods now: sentinels and servers. As before, they both use the same docker image. Sentinels are still looking for an environment variable to tell them to specialize. During initial launch, the server pods each perform a deterministic check to determine which of them should become the master. Once the master is up, the sentinels connect to it and they assign roles to to remaining slaves.

The sentinels move servers in/out of the master/replica LBs as needed using the new promote.sh script. This way individual redis servers can fail and be replaced without data loss or downtime beyond what it took to reach +odown. 

Multiple parallel executions of the chart in the same K8S cluster are explicitly supported. There are a few new hooks to make customizing and debug a bit easier. Labels such as podID and runID are set for each server on launch. Server promotion and initial label setting have their own scripts.

**Comment by ianmaddox** [comment]
Bumped chart major version.

File: stable/redis-ha/templates/redis-slave-statefulset.yaml (hunk 7 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-slave-statefulset.yaml:1]: This functionality was merged into the master pod type to create the server pod type.

**Comment by bzumhagen** [comment]
Just chiming in to say this looks good, and I'm eagerly waiting this merge.

File: stable/redis-ha/values.yaml (hunk 6 above)
Comment (ianmaddox) [reply stable/redis-ha/values.yaml]: Full changeset is here:
https://github.com/smileisak/docker-images/pull/4

File: stable/redis-ha/templates/redis-master-appliance-vpc.yaml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "fullname" . }}-master-appliance-vpc
+  annotations:
+    cloud.google.com/load-balancer-type: "Internal"
```
Comment (viglesiasce) [inline stable/redis-ha/templates/redis-master-appliance-vpc.yaml]: Can we factor this out to the values file?

Here is a good example:
https://github.com/kubernetes/charts/blob/master/stable/prometheus/values.yaml#L139
https://github.com/kubernetes/charts/blob/master/stable/prometheus/templates/server-service.yaml#L6

File: stable/redis-ha/templates/redis-master-appliance-vpc.yaml [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  ports:
+  - port: 6379
+    protocol: TCP
+    targetPort: 6379
+  selector:
+    app: "redis-ha"
+    redis-node: "true"
+    redis-role: "master"
+    release: "{{ .Release.Name }}"
+  type: LoadBalancer
```
Comment (viglesiasce) [inline stable/redis-ha/templates/redis-master-appliance-vpc.yaml]: Make this a ClusterIP by default to prevent unwanted exposure OOTB.

File: stable/redis-ha/templates/redis-slave-appliance-vpc.yaml [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  ports:
+  - port: 6379
+    protocol: TCP
+    targetPort: 6379
+  selector:
+    app: "redis-ha"
+    redis-node: "true"
+    redis-role: "slave"
+    release: "{{ .Release.Name }}"
+  type: LoadBalancer
```
Comment (viglesiasce) [inline stable/redis-ha/templates/redis-slave-appliance-vpc.yaml]: Same as above RE: ClusterIP

File: stable/redis-ha/templates/redis-slave-appliance-vpc.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "fullname" . }}-slave-appliance-vpc
+  annotations:
```
Comment (viglesiasce) [inline stable/redis-ha/templates/redis-slave-appliance-vpc.yaml]: Same as above RE: factoring out to values

**Review by viglesiasce — CHANGES_REQUESTED** [review]
Good stuff @ianmaddox just two nits on the defaults for the services and the annotations.

File: stable/redis-ha/templates/redis-slave-statefulset.yaml (hunk 7 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-slave-statefulset.yaml:1]: Statefulsets stop auto-scaling in K8S when the first instance is orphaned. This causes problems when SS instances are re-labeled on the fly to move between load balancers. The compromise was to fall back to dynamic naming and use logic instead of hostname to determine which box should be the first master.

More info on the SS issue:
https://github.com/kubernetes/kubernetes/issues/46901

File: stable/redis-ha/templates/redis-master-appliance-vpc.yaml (hunk 9 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-master-appliance-vpc.yaml]: Done. Added appliance.vpcMode flag to control this. Defaults to ClusterIP.

File: stable/redis-ha/templates/redis-slave-appliance-vpc.yaml (hunk 11 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-slave-appliance-vpc.yaml]: Done.

File: stable/redis-ha/templates/redis-slave-appliance-vpc.yaml (hunk 10 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-slave-appliance-vpc.yaml]: Done

File: stable/redis-ha/templates/redis-master-appliance-vpc.yaml (hunk 8 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-master-appliance-vpc.yaml]: Done.

File: stable/redis-ha/templates/redis-master-appliance-vpc.yaml [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  ports:
+  - port: 6379
+    protocol: TCP
+    targetPort: 6379
+  selector:
+    app: "redis-ha"
+    redis-node: "true"
+    redis-role: "master"
+    release: "{{ .Release.Name }}"
+{{- if eq .Values.appliance.vpcMode true }}
```
Comment (viglesiasce) [inline stable/redis-ha/templates/redis-master-appliance-vpc.yaml]: Would prefer this just value in the values file that is defaulted to ClusterIP. For example:
https://github.com/kubernetes/charts/blob/master/stable/prometheus/values.yaml#L156

File: stable/redis-ha/templates/redis-pvc.yaml [hunk 13]
```diff
@@ -3,7 +3,7 @@
 kind: PersistentVolumeClaim
 apiVersion: v1
 metadata:
-  name: {{ template "fullname" . }}-slave
+  name: {{ template "fullname" . }}-pvc
```
Comment (viglesiasce) [inline stable/redis-ha/templates/redis-pvc.yaml]: Since the type is already know to be a PVC it is more indicative to leave a more descriptive name of what the PVC is for.

File: stable/redis-ha/templates/redis-server-deployment.yaml [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+            value: {{ template "fullname" . }}-
+        ports:
+          - containerPort: 6379
+        volumeMounts:
+          - mountPath: /redis-master-data 
+            name: data
+      volumes:
+      - name: data
+        emptyDir: {}
+#        persistentVolumeClaim:
```
Comment (viglesiasce) [inline stable/redis-ha/templates/redis-server-deployment.yaml]: Is it the case that this should _never_ have storage attached? If so remove these lines and the mounted emptydir volume for data. If so we should change this to a stateful set with a PVC template.

File: stable/redis-ha/templates/redis-slave-appliance-vpc.yaml [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+    targetPort: 6379
+  selector:
+    app: "redis-ha"
+    redis-node: "true"
+    redis-role: "slave"
+    release: "{{ .Release.Name }}"
+{{- if eq .Values.appliance.vpcMode true }}
+  type: LoadBalancer
+{{- else }}
+  type: ClusterIP
```
Comment (viglesiasce) [inline stable/redis-ha/templates/redis-slave-appliance-vpc.yaml]: Same as above please make this a configurable value defaulted to ClusterIP.

File: stable/redis-ha/values.yaml [hunk 16]
```diff
@@ -66,9 +60,13 @@ persistentVolume:
...
-  master: 1
-  slave: 1
-  sentinel: 3
+  servers: 3
+  sentinels: 3
+appliance:
+  # Set vpcMode to true to automatically open the master and slave load balancers to access from within the VPC
+  vpcMode: false
+  annotations:
+    cloud.google.com/load-balancer-type: "Internal"
```
Comment (viglesiasce) [inline stable/redis-ha/values.yaml]: Default the annotations to empty as we prefer to not keep cloud provider specific info in the chart if not needed out of the box.

File: stable/redis-ha/values.yaml [hunk 17]
```diff
@@ -66,9 +60,13 @@ persistentVolume:
...
-redis_image: quay.io/smile/redis:4.0.2
+redis_image: quay.io/ianmaddox/redis-sentinel:redis-4.0.6r0
 ## replicas number for each component
 replicas:
-  master: 1
-  slave: 1
-  sentinel: 3
+  servers: 3
+  sentinels: 3
+appliance:
```
Comment (viglesiasce) [inline stable/redis-ha/values.yaml]: What is meant by appliance here? Can you add some info in the readme about how/when to configure like this?

File: stable/redis-ha/values.yaml [hunk 18]
```diff
@@ -66,9 +60,13 @@ persistentVolume:
...
+  # Set vpcMode to true to automatically open the master and slave load balancers to access from within the VPC
+  vpcMode: false
+  annotations:
```
Comment (viglesiasce) [inline stable/redis-ha/values.yaml]: This applies to the service for a particular component. Can you add the configuration in that section of the values file?

**Review by viglesiasce — CHANGES_REQUESTED** [review]

File: stable/redis-ha/values.yaml (hunk 17 above)
Comment (ianmaddox) [reply stable/redis-ha/values.yaml]: Good catch. I put all the documentation in the Dockerfile for the image. Added helpful notes to README.

File: stable/redis-ha/templates/redis-server-deployment.yaml (hunk 14 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-server-deployment.yaml]: Cleaned this up. We cannot use statefulsets because of the orphaned root node issue mentioned above.

File: stable/redis-ha/templates/redis-pvc.yaml (hunk 13 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-pvc.yaml]: Donezo.

File: stable/redis-ha/templates/redis-server-deployment.yaml (hunk 14 above)
Comment (ianmaddox) [reply stable/redis-ha/templates/redis-server-deployment.yaml]: Removed all references to persistence.

**Comment by frekw** [comment]
@ianmaddox perhaps you could have a look at fixing https://github.com/smileisak/docker-images/issues/5 while you're at it since you're basically rewriting the image. Should be a simple fix.

**Review by viglesiasce — APPROVED** [review]

**Comment by viglesiasce** [comment]
@frekw as noted in that issue, you will need to change that value on your host not in the container image.

**Comment by viglesiasce** [comment]
Failover worked like a charm. Thanks @ianmaddox!!!

/lgtm

**Comment by frekw** [comment]
@viglesiasce ah, fair enough!

**Comment by frekw** [comment]
The solution seems to be an init container (as per https://stackoverflow.com/a/45283978). Should that be added as part of the deployment? I'm not well versed enough in the linux kernel to understand what potential negative effects it could have on other pods schedules on the same node

**Comment by patricklx** [comment]
@ianmaddox Can you explain why PVC was removed?
Is there a specific reason or can we just enable it by copying the chart and adapt it?
